From 65ab5e074960b5bde41cfeb7c51984470db46064 Mon Sep 17 00:00:00 2001
From: Ken Ma <make@marvell.com>
Date: Wed, 16 Mar 2016 21:12:40 +0800
Subject: [PATCH 294/538] arm64: neta: a3700: add SFP support both in 1G/2.5G
 modes

- Add a new dts for SFP 1G/2.5G board configuration, keep neta0 in
  RGMII mode and set neta1 to SFP over SGMII1 in the fixed-link mode;
- Now for the fixed-link mode, speed setting and duplex setting are
  supported while pause setting for flow control is not supported yet.

Change-Id: If6411f4ac4418ee708f74db065b42c1eff74c3ab
Signed-off-by: Ken Ma <make@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/28279
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Victor Gu <xigu@marvell.com>
Reviewed-by: Wilson Ding <dingwei@marvell.com>
---
 .../bindings/net/marvell-armada-370-neta.txt       |  3 +-
 arch/arm64/boot/dts/marvell/Makefile               |  1 +
 arch/arm64/boot/dts/marvell/armada-3720-db-sfp.dts | 70 ++++++++++++++++++++++
 3 files changed, 73 insertions(+), 1 deletion(-)
 create mode 100644 arch/arm64/boot/dts/marvell/armada-3720-db-sfp.dts

diff --git a/Documentation/devicetree/bindings/net/marvell-armada-370-neta.txt b/Documentation/devicetree/bindings/net/marvell-armada-370-neta.txt
index d2cb856..93c9e85 100644
--- a/Documentation/devicetree/bindings/net/marvell-armada-370-neta.txt
+++ b/Documentation/devicetree/bindings/net/marvell-armada-370-neta.txt
@@ -7,11 +7,12 @@ Required properties:
 	"marvell,armada3700-neta"
 - reg: address and length of the register set for the device.
 - interrupts: interrupt for the device
-- phy: See ethernet.txt file in the same directory.
 - phy-mode: See ethernet.txt file in the same directory
 - clocks: a pointer to the reference clock for this device.
 
 Optional properties:
+- phy: See ethernet.txt file in the same directory.
+- fixed-link: See fixed-link.txt file in the same directory.
 - tx-csum-limit: maximum mtu supported by port that allow TX checksum.
   Value is presented in bytes. If not used, by default 1600B is set for
   "marvell,armada-370-neta" and 9800B for others.
diff --git a/arch/arm64/boot/dts/marvell/Makefile b/arch/arm64/boot/dts/marvell/Makefile
index fed8e5d..628a026 100644
--- a/arch/arm64/boot/dts/marvell/Makefile
+++ b/arch/arm64/boot/dts/marvell/Makefile
@@ -5,6 +5,7 @@ dtb-$(CONFIG_ARCH_BERLIN) += berlin4ct-stb.dtb
 # Mvebu SoC Family
 dtb-$(CONFIG_ARCH_MVEBU) += armada-3720-db.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-3720-db-sgmii1.dtb
+dtb-$(CONFIG_ARCH_MVEBU) += armada-3720-db-sfp.dtb
 dtb-$(CONFIG_ARCH_MVEBU) += armada-7040-db-router.dtb
 # For Armada-7040-Z1 revision
 dtb-$(CONFIG_ARCH_MVEBU) += armada-7040-rz-db-router.dtb
diff --git a/arch/arm64/boot/dts/marvell/armada-3720-db-sfp.dts b/arch/arm64/boot/dts/marvell/armada-3720-db-sfp.dts
new file mode 100644
index 0000000..e7827dd
--- /dev/null
+++ b/arch/arm64/boot/dts/marvell/armada-3720-db-sfp.dts
@@ -0,0 +1,70 @@
+/*
+ * Device Tree file for Marvell Armada 3720 development board
+ * (DB-88F3720-DDR3)
+ * Copyright (C) 2016 Marvell
+ *
+ * Ken Ma <make@marvell.com>
+ *
+ * This file is dual-licensed: you can use it either under the terms
+ * of the GPL or the X11 license, at your option. Note that this dual
+ * licensing only applies to this file, and not this project as a
+ * whole.
+ *
+ *  a) This file is free software; you can redistribute it and/or
+ *     modify it under the terms of the GNU General Public License as
+ *     published by the Free Software Foundation; either version 2 of the
+ *     License, or (at your option) any later version.
+ *
+ *     This file is distributed in the hope that it will be useful
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ * Or, alternatively
+ *
+ *  b) Permission is hereby granted, free of charge, to any person
+ *     obtaining a copy of this software and associated documentation
+ *     files (the "Software"), to deal in the Software without
+ *     restriction, including without limitation the rights to use
+ *     copy, modify, merge, publish, distribute, sublicense, and/or
+ *     sell copies of the Software, and to permit persons to whom the
+ *     Software is furnished to do so, subject to the following
+ *     conditions:
+ *
+ *     The above copyright notice and this permission notice shall be
+ *     included in all copies or substantial portions of the Software.
+ *
+ *     THE SOFTWARE IS PROVIDED , WITHOUT WARRANTY OF ANY KIND
+ *     EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
+ *     OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ *     NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ *     HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY
+ *     WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ *     FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ *     OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "armada-3720-db.dts"
+/* This file enables eth1 to work in SGMII mode with 1g or 2.5g fixed-link,
+ * eth0 still works in RGMII mode.
+ */
+/ {
+	soc {
+		internal-regs {
+			eth1: ethernet@40000 {
+				phy-mode = "sgmii";
+				status = "okay";
+				fixed-link {
+					/* Here "speed" is set to 1000, GBE MAC is running in 1G mode, */
+					/* but the actual MAC speed may be 1 Gbps or 2.5 Gbps, */
+					/* it's up to the corresponding SERDES speed in u-boot fdt comphy node. */
+					/* If SERDES speed is set to 3.125G in u-boot, it implies sgmii 2.5 Gbps; */
+					/* if SERDES speed is set to 1.25G in u-boot, it implies sgmii 1 Gbps. */
+					speed = <1000>;
+					full-duplex;
+				};
+			};
+		};
+	};
+};
+
-- 
1.9.1

